lib/phlex/sgml.rb

```diff
@@ -491,6 +491,45 @@ def __nested_attributes__(attributes, base_name, buffer = +"")
 			end
 		end
 
+		# @api private
+		def __nested_tokens__(tokens)
+			buffer = +""
+
+			i, length = 0, tokens.length
+
+			while i < length
+				token = tokens[i]
+
+				case token
+				when nil
+					next
+				when String
+					if i > 0
+						buffer << " " << token
+					else
+						buffer << token
+					end
+				when Symbol
+					if i > 0
+						buffer << " " << token.name
+					else
+						buffer << token.name
+					end
```

Comment:
I spent too long in a `bench.rb` file trying to beat this. But I think you've already wrung out all the perf that can be gained here. I'm frustrated that all these `if i > 0` checks are faster than always including the space and using `delete_suffix!(" ")` or `delete_prefix!(" ")` (depending on where you put the space).

That being said, I think for _most_ token lists, moving the `nil` check to the end will be faster. Also there is a bug where `i` never gets incremented when you encounter a nil.

```suggestion
				when String
					if i > 0
						buffer << " " << token
					else
						buffer << token
					end
				when Symbol
					if i > 0
						buffer << " " << token.name
					else
						buffer << token.name
					end
				when nil
```